File: Assets/Bundles/Shared/Resources/TS/UI/ProgressBarGraphics.lua
-- Compiled with unity-ts v2.1.0-75
local Task = require("Shared/TS/Util/Task").Task
local GameObjectBridge = require("Shared/TS/GameObjectBridge").GameObjectBridge
local ProgressBarGraphics
do
	ProgressBarGraphics = setmetatable({}, {
		__tostring = function()
			return "ProgressBarGraphics"
		end,
	})
	ProgressBarGraphics.__index = ProgressBarGraphics
	function ProgressBarGraphics.new(...)
		local self = setmetatable({}, ProgressBarGraphics)
		return self:constructor(...) or self
	end
	function ProgressBarGraphics:constructor(transform, options)
		self.TransformKey = "Transforms"
		self.GraphicsKey = "Graphics"
		self.AnimKey = "Animations"
		self.fillDurationInSeconds = 0.08
		self.changeDelayInSeconds = 0.3
		self.changeDurationInSeconds = 0.125
		self.enabled = true
		self.deathOnZero = true
		self.currentDelta = 0
		self.transform = transform.gameObject:GetComponent("RectTransform")
		self.refs = transform:GetComponent("GameObjectReferences")
		self.fillImage = self.refs:GetValue(self.GraphicsKey, "Fill")
		self.fillTransform = self.refs:GetValue(self.TransformKey, "Fill")
		self.changeFillTransform = self.refs:GetValue(self.TransformKey, "ChangeFill")
		self.growthFillTransform = self.refs:GetValue(self.TransformKey, "GrowthFill")
		self.graphicsHolder = self.refs:GetValue(self.TransformKey, "GraphicsHolder")
		self.brokenGraphicsHolder = self.refs:GetValue(self.TransformKey, "BrokenGraphicsHolder")
		self.deathAnim = self.refs:GetValue(self.AnimKey, "Finished")
		self.graphicsHolder.gameObject:SetActive(true)
		self.brokenGraphicsHolder.gameObject:SetActive(false)
		local _result = options
		if _result ~= nil then
			_result = _result.deathOnZero
		end
		local _condition = _result
		if _condition == nil then
			_condition = true
		end
		self.deathOnZero = _condition
		local _result_1 = options
		if _result_1 ~= nil then
			_result_1 = _result_1.fillColor
		end
		if _result_1 then
			self:SetColor(options.fillColor)
		end
		local _fn = self
		local _result_2 = options
		if _result_2 ~= nil then
			_result_2 = _result_2.initialPercentDelta
		end
		local _condition_1 = _result_2
		if _condition_1 == nil then
			_condition_1 = 1
		end
		_fn:InstantlySetValue(_condition_1)
		self.enabled = true
	end
	function ProgressBarGraphics:OnDelete()
		self.fillTransform:TweenCancelAll(false, true)
		self.changeFillTransform:TweenCancelAll(false, true)
		self.enabled = false
		GameObjectBridge:Destroy(self.transform.gameObject)
	end
	function ProgressBarGraphics:SetColor(newColor)
		self.fillImage.color = newColor
	end
	function ProgressBarGraphics:InstantlySetValue(percentDelta)
		self.currentDelta = percentDelta
		local fillScale = Vector3.new(percentDelta, self.fillTransform.localScale.y, self.fillTransform.localScale.z)
		self.fillTransform.localScale = fillScale
		self.changeFillTransform.localScale = fillScale
	end
	function ProgressBarGraphics:SetValue(percentDelta)
		if self.deathOnZero and percentDelta <= 0 then
			-- Wait for the change animation
			Task:Delay(self.fillDurationInSeconds, function()
				if self.transform then
					-- Play the death animation
					self.deathAnim:Play()
					self.graphicsHolder.gameObject:SetActive(false)
					self.brokenGraphicsHolder.gameObject:SetActive(true)
					Task:Delay(1.1, function()
						if self.transform and self.currentDelta > 0 then
							-- Reset if the progress has filled back up (Respawn)
							self:SetValue(self.currentDelta)
						end
					end)
				end
			end)
		else
			self.deathAnim:Stop()
			self.graphicsHolder.gameObject:SetActive(true)
			self.brokenGraphicsHolder.gameObject:SetActive(false)
		end
		-- Animate fill down
		self.fillTransform:TweenLocalScaleX(percentDelta, self.fillDurationInSeconds)
		if percentDelta > self.currentDelta then
			-- Growth
			self.changeFillTransform.gameObject:SetActive(false)
			self.growthFillTransform.gameObject:SetActive(true)
			self.growthFillTransform.localScale = Vector3.new(percentDelta - self.currentDelta, self.growthFillTransform.localScale.y, self.growthFillTransform.localScale.z)
			self.growthFillTransform.anchoredPosition = Bridge:MakeVector2(self.transform.rect.width * self.currentDelta, self.growthFillTransform.anchoredPosition.y)
			self.growthFillTransform:TweenLocalScaleX(0, self.fillDurationInSeconds)
			self.growthFillTransform:TweenAnchoredPositionX(self.transform.rect.width * percentDelta, self.changeDurationInSeconds)
		else
			-- Decay
			self.growthFillTransform.gameObject:SetActive(false)
			self.changeFillTransform.gameObject:SetActive(true)
			-- Hold then animate change indicator
			Task:Delay(self.changeDelayInSeconds, function()
				if not self.enabled then
					return nil
				end
				self.changeFillTransform:TweenLocalScaleX(percentDelta, self.changeDurationInSeconds)
			end)
		end
		self.currentDelta = percentDelta
	end
end
return {
	ProgressBarGraphics = ProgressBarGraphics,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Shared/UI/ProgressBarGraphics.ts
import { Task } from "../Util/Task";
import { GameObjectBridge } from "../GameObjectBridge";

export interface ProgressBarOptions {
	initialPercentDelta?: number;
	fillColor?: Color;
	deathOnZero?: boolean;
}

export class ProgressBarGraphics {
	private readonly TransformKey = "Transforms";
	private readonly GraphicsKey = "Graphics";
	private readonly AnimKey = "Animations";
	private transform: RectTransform;
	private refs: GameObjectReferences;
	private fillImage: Image;
	private fillTransform: RectTransform;
	private changeFillTransform: RectTransform;
	private growthFillTransform: RectTransform;
	private graphicsHolder: RectTransform;
	private brokenGraphicsHolder: RectTransform;
	private deathAnim: Animation;

	public fillDurationInSeconds = 0.08;
	public changeDelayInSeconds = 0.3;
	public changeDurationInSeconds = 0.125;
	private enabled = true;
	private deathOnZero = true;
	private currentDelta = 0;

	constructor(transform: Transform, options?: ProgressBarOptions) {
		this.transform = transform.gameObject.GetComponent<RectTransform>();
		this.refs = transform.GetComponent<GameObjectReferences>();
		this.fillImage = this.refs.GetValue<Image>(this.GraphicsKey, "Fill");
		this.fillTransform = this.refs.GetValue<RectTransform>(this.TransformKey, "Fill");
		this.changeFillTransform = this.refs.GetValue<RectTransform>(this.TransformKey, "ChangeFill");
		this.growthFillTransform = this.refs.GetValue<RectTransform>(this.TransformKey, "GrowthFill");
		this.graphicsHolder = this.refs.GetValue<RectTransform>(this.TransformKey, "GraphicsHolder");
		this.brokenGraphicsHolder = this.refs.GetValue<RectTransform>(this.TransformKey, "BrokenGraphicsHolder");
		this.deathAnim = this.refs.GetValue<Animation>(this.AnimKey, "Finished");

		this.graphicsHolder.gameObject.SetActive(true);
		this.brokenGraphicsHolder.gameObject.SetActive(false);

		this.deathOnZero = options?.deathOnZero ?? true;

		if (options?.fillColor) {
			this.SetColor(options.fillColor);
		}
		this.InstantlySetValue(options?.initialPercentDelta ?? 1);
		this.enabled = true;
	}

	public OnDelete() {
		this.fillTransform.TweenCancelAll(false, true);
		this.changeFillTransform.TweenCancelAll(false, true);
		this.enabled = false;
		GameObjectBridge.Destroy(this.transform.gameObject);
	}

	public SetColor(newColor: Color) {
		this.fillImage.color = newColor;
	}

	public InstantlySetValue(percentDelta: number) {
		this.currentDelta = percentDelta;
		let fillScale = new Vector3(percentDelta, this.fillTransform.localScale.y, this.fillTransform.localScale.z);
		this.fillTransform.localScale = fillScale;
		this.changeFillTransform.localScale = fillScale;
	}

	public SetValue(percentDelta: number) {
		if (this.deathOnZero && percentDelta <= 0) {
			//Wait for the change animation
			Task.Delay(this.fillDurationInSeconds, () => {
				if (this.transform) {
					//Play the death animation
					this.deathAnim.Play();
					this.graphicsHolder.gameObject.SetActive(false);
					this.brokenGraphicsHolder.gameObject.SetActive(true);
					Task.Delay(1.1, () => {
						if (this.transform && this.currentDelta > 0) {
							//Reset if the progress has filled back up (Respawn)
							this.SetValue(this.currentDelta);
						}
					});
				}
			});
		} else {
			this.deathAnim.Stop();
			this.graphicsHolder.gameObject.SetActive(true);
			this.brokenGraphicsHolder.gameObject.SetActive(false);
		}

		//Animate fill down
		this.fillTransform.TweenLocalScaleX(percentDelta, this.fillDurationInSeconds);

		if (percentDelta > this.currentDelta) {
			//Growth
			this.changeFillTransform.gameObject.SetActive(false);
			this.growthFillTransform.gameObject.SetActive(true);
			this.growthFillTransform.localScale = new Vector3(
				percentDelta - this.currentDelta,
				this.growthFillTransform.localScale.y,
				this.growthFillTransform.localScale.z,
			);
			this.growthFillTransform.anchoredPosition = Bridge.MakeVector2(
				this.transform.rect.width * this.currentDelta,
				this.growthFillTransform.anchoredPosition.y,
			);

			this.growthFillTransform.TweenLocalScaleX(0, this.fillDurationInSeconds);
			this.growthFillTransform.TweenAnchoredPositionX(
				this.transform.rect.width * percentDelta,
				this.changeDurationInSeconds,
			);
		} else {
			//Decay
			this.growthFillTransform.gameObject.SetActive(false);
			this.changeFillTransform.gameObject.SetActive(true);

			//Hold then animate change indicator
			Task.Delay(this.changeDelayInSeconds, () => {
				if (!this.enabled) return;
				this.changeFillTransform.TweenLocalScaleX(percentDelta, this.changeDurationInSeconds);
			});
		}

		this.currentDelta = percentDelta;
	}
}
